crates/autopilot/src/solvable_orders.rs

```diff
@@ -257,22 +258,44 @@ impl SolvableOrdersCache {
         prices: &BTreeMap<H160, U256>,
     ) -> Vec<Order> {
         orders.retain(|order| {
-            if let OrderClass::Limit(limit) = &order.metadata.class {
-                // Convert the sell and buy price to the native token (ETH) and make sure that sell
-                // with the surplus fee is higher than buy with the configurable price factor.
-                let sell_native = (order.data.sell_amount + limit.surplus_fee)
-                    * prices.get(&order.data.sell_token).unwrap();
-                let buy_native = order.data.buy_amount * prices.get(&order.data.buy_token).unwrap();
-                let sell_native = u256_to_big_decimal(&sell_native);
-                let buy_native = u256_to_big_decimal(&buy_native);
-                if sell_native >= buy_native * self.limit_order_price_factor.clone() {
-                    true
-                } else {
-                    tracing::debug!(order_uid = %order.metadata.uid, "limit order is outside market price, skipping");
-                    false
+            let surplus_fee = match &order.metadata.class {
+                OrderClass::Limit(limit) => limit.surplus_fee,
+                _ => return true,
+            };
+
+            let sell_price = *prices.get(&order.data.sell_token).unwrap();
+            let buy_price = *prices.get(&order.data.buy_token).unwrap();
+
+            // Convert the sell and buy price to the native token (ETH) and make sure that sell
+            // discounting the surplus fee is higher than buy with the configurable price factor.
+            let (sell_native, buy_native) = match (
+                order
+                    .data
+                    .sell_amount
+                    .checked_sub(surplus_fee)
+                    .and_then(|sell| sell.checked_mul(sell_price)),
+                order.data.buy_amount.checked_mul(buy_price),
+            ) {
+                (Some(sell), Some(buy)) => (sell, buy),
+                _ => {
+                    tracing::debug!(
+                        order_uid = %order.metadata.uid,
+                        "limit order overflow computing native amounts; skipping",
+                    );
+                    return false;
                 }
```

Comment:
First, I thought this was just a refactor. After, I saw that you changed the `sell_amount` and `surplus_fee` to being deducted.
Was this intentional? Do we have a unit test for this function? (seems like a decent candidate for one)

Reply:
Ah sorry, now I saw the PR description. Anyway, I still think it would be nice to add a unit test for `filter_mispriced_limit_orders`

Reply:
Good point, will do.

Reply:
Added unit tests.